pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,469 @@
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	latencyConfig := LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	m.latencyConfigChanged <- latencyConfig
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	klog.V(4).InfoS("NodeLatencyMonitor deleted")
+	latencyConfig := LatencyConfig{Enable: false}
+
+	m.latencyConfigChanged <- latencyConfig
+}
+
+// sendPing sends an ICMP message to the target IP address.
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	ip := &net.IPAddr{IP: addr}
+
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   int(icmpEchoID),
+		Seq:  int(seqID),
+		Data: []byte(timeStart.Format(time.RFC3339Nano)),
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.V(4).InfoS("Sending ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+// recvPing receives an ICMP message from the target IP address.
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool) {
+	// We only expect small packets, if we receive a larger packet, we will drop the extra data.
+	readBuffer := make([]byte, 128)
+	for {
+		n, peer, err := socket.ReadFrom(readBuffer)
+		if err != nil {
+			// When the socket is closed in the Run method, this error will be logged, which is not ideal.
+			// In the future, we may try setting a ReadDeadline on the socket before each ReadFrom and using
+			// a channel to signal that the loop should terminate.
+			klog.ErrorS(err, "Failed to read ICMP message")
+			return
+		}
+
+		destIP := peer.String()
+
+		// Parse the ICMP message
+		var msg *icmp.Message
+		if isIPv4 {
+			msg, err = icmp.ParseMessage(protocolICMP, readBuffer[:n])
+			if err != nil {
+				klog.ErrorS(err, "Failed to parse ICMP message")
+				continue
+			}
+			if msg.Type != ipv4.ICMPTypeEcho && msg.Type != ipv4.ICMPTypeEchoReply {
+				klog.V(5).InfoS("Ignoring non-ping ICMP message", "msg", msg)
+				continue
+			}
+			// Ignore ICMP echo messages received from other Nodes (they will be answered by the system)
+			if msg.Type == ipv4.ICMPTypeEcho {
+				klog.V(7).InfoS("Ignoring ICMP echo request message", "msg", msg)
+				continue
+			}
+		} else {
+			msg, err = icmp.ParseMessage(protocolICMPv6, readBuffer)
```

Comment:
should it parse `readBuffer[:n]` only?

Reply:
Yes, [:n] can be selected for consistency with ipv4 case.